Update first-time contributors per Volto Team discussion

stevepiercy opened the pull request:
This update sets expectations that were "just known", but are not yet communicated in writing. We should also ask the Classic UI Team and Steering Circle to review these changes, as well as the entire page.

See the [preview](https://deploy-preview-1556--6-docs-plone-org.netlify.app/contributing/first-time.html).

stevepiercy committed:
Update first-time contributors per Volto Team discussion

tkimnguyen commented:
This is great! I lol'd at some of the stuff you've replaced 😝🥳

tkimnguyen commented:
Could we add something about coming to a conference sprint if they want an easy way to onboard?

stevepiercy committed:
Add mention of participating in an event for support, training, or mentoring

stevepiercy commented:
> Could we add something about coming to a conference sprint if they want an easy way to onboard?

Done in bdeff45f217dcd61922e8c1241ebea7fddcd2978

stevepiercy committed:
Merge branch '6.0' into first-time-contributor-update

sneridagh commented:
@stevepiercy I'd add also to the first paragraph being actively working on a Plone project. Even if it's a hobby one. Although nobody will learn all the above, just because... 

Also, explaining the label differences, starting with PLIPs, so they are suited for seasoned Plone developers and probably imply "weeks" of work.

And, yes, we should put more effort in issues and PRs label herding.

stevepiercy committed:
Add mention of working on a Plone project as a requirement, and how using filters for finding appropriate issues to work on

stevepiercy commented:
@sneridagh done in 74c240a00bbc54d2fd7408588d0289263679124e

stevepiercy committed:
If you installed Plone, you have tried it, thus this expectation is redundant

stevepiercy committed:
Slight rewording for clarification, per feedback from @fredvd

fredvd reviewed (approved):
Thank you, looks good!

davisagli committed:
Merge branch '6.0' into first-time-contributor-update

davisagli commented on docs/contributing/first-time.md:
```diff
@@ -168,15 +197,26 @@ Once you have completed, tested, and linted your code, and created a {ref}`contr
 
 1.  Visit your fork of the Plone repository on GitHub, and [create a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) against the development branch.
     -   Make your title and description descriptive.
```
Suggestion for clarity:

```suggestion
    -   Make your title and description descriptive.
        Remember that the reviewers are looking at many pull requests and need to be able to understand the context quickly.
        A lazy title like "Fixes bug" is meaningless and will get lost.
```

stevepiercy replied:
I combined and shortened the three suggestions for this section into dc24c323192cda1b89817686b63460bd73cee4d2

davisagli commented on docs/contributing/first-time.md:
```diff
@@ -168,15 +197,26 @@ Once you have completed, tested, and linted your code, and created a {ref}`contr
     -   Make your title and description descriptive.
-        Don't be lazy with "Fixes bug" or other useless drivel.
-    -   To automatically close the original issue when your pull request is merged, include "Closes #issue_number" in your description.
+    -   Include "Fixes #" and the related issue number.
```
```suggestion
    -   Include "Fixes #" and the related issue number in the description.
```

davisagli commented on docs/contributing/first-time.md:
```diff
@@ -168,15 +197,26 @@ Once you have completed, tested, and linted your code, and created a {ref}`contr
+    -   Include "Fixes #" and the related issue number.
+        This enables automatic closing of the related issue when the pull request is merged.
         This also creates a hyperlink to the original issue for easy reference.
+    -   Don't be lazy with "Fixes bug" or other useless drivel.
```
This isn't very clear, I suggested an alternative above.
```suggestion
```

davisagli commented on docs/contributing/first-time.md:
```diff
@@ -168,15 +197,26 @@ Once you have completed, tested, and linted your code, and created a {ref}`contr
-Provide instructions when the contributor needs to update their pull request with changes from the default branch.
-Members of Contributors do not have the button "Update branch" to easily do this, and must use git foo to manage the situation.
+
+### Update your pull request from your fork
```
I don't think these instructions will do the right thing when working from a fork. It will pull the main branch from the fork, but not update it from the upstream repository.

stevepiercy replied:
Oh, that's right, unless we track the upstream, which we should do, too. Pull from upstream, push to origin. I forgot that, being a privileged member.

stevepiercy replied:
@davisagli I made a couple of revisions. Would you please take another look to verify my sad gitfu?

davisagli replied:
@stevepiercy That's better. I'm making a couple more suggestions.

stevepiercy committed:
Improve how to write a good pull request title and description

stevepiercy committed:
Add basic Plone git workflow narrative text and diagram.

stevepiercy committed:
Swap narrative explanation and diagram with steps to set up an environment

davisagli commented on docs/contributing/first-time.md:
```diff
@@ -167,16 +210,28 @@ Once you have completed, tested, and linted your code, and created a {ref}`contr
+### Update your pull request from your fork
+
+Often another pull request will get merged before yours, and your pull request will fall behind the main branch.
+To keep your pull request current, you can issue the follow git commands.
```
```suggestion
To keep your pull request current, you can issue the following git commands.
```

stevepiercy committed:
Alter git commands to track upstream and push to origin

davisagli commented on docs/contributing/first-time.md:
````diff
@@ -163,20 +208,32 @@ Once you have completed, tested, and linted your code, and created a {ref}`contr
+
+```shell
+# Assume `main` is the main branch.
+git checkout main
+git pull
+git checkout my-branch-name
+git merge main
+git push origin my-branch-name
````
I think instead of these commands it's simpler to suggest:
```
# (Assuming you are on the branch for your PR which you want to update)
git fetch upstream  # assumes the upstream remote was configured based on https://docs.github.com/en/get-started/quickstart/fork-a-repo#configuring-git-to-sync-your-fork-with-the-upstream-repository
git merge upstream/main  # You might need to resolve conflicts here.
git push
```

stevepiercy replied:
I think I got it now, for this suggestion and the next, in 2ec39f33fadf9dd521c4dde5a8dede782041578b. One more review, please?

davisagli commented on docs/contributing/first-time.md:
```diff
@@ -163,20 +208,32 @@ Once you have completed, tested, and linted your code, and created a {ref}`contr
     git add <files>
     git commit -m "My commit message"
-    git push
+    git push origin my-branch-name
```
```suggestion
    git push -u origin my-branch-name
```

This sets origin/my-branch-name as the upstream (remote) branch for this local branch, so that in the future you can run `git push` on this branch and it will know where to push it.

davisagli replied:
@stevepiercy We still need this change adding `-u` so that the `git push` below knows which branch to push to.

stevepiercy replied:
Oops, went too fast. Suggestion committed.

stevepiercy committed:
Update docs/contributing/first-time.md

Co-authored-by: David Glick <david@glicksoftware.com>

stevepiercy committed:
Polish up "Update your pull request from your fork" section

stevepiercy committed:
Update docs/contributing/first-time.md

Co-authored-by: David Glick <david@glicksoftware.com>

davisagli reviewed (dismissed):
Okay, I think these instructions at least have a chance of working now. :)

stevepiercy commented:
I'm going to wait for the Steering Circle meeting next week to give them all a chance to provide feedback.

MrTango commented on docs/contributing/first-time.md:
```diff
@@ -106,6 +133,20 @@ After you have satisfied the above steps and have clear direction on how to proc
 
 ### Set up your environment
 
+As a member of the Plone Contributors Team, you do not have write access to push commits to GitHub repositories under the Plone organization.
```
is this really the case?
Normally you could create a branch, work on it and create a PR.
Or does this require higher levels of permissions?

stevepiercy replied:
Yes, this is the case for first-time contributors. It was a severe security issue, and created a lot of other problems.

When GitHub was friendly, and we used the paper-based Plone Contributor Agreement, we tossed everyone into the [Developers Team](https://github.com/orgs/plone/teams/developers). Then we got hammered by Hacktoberfest and GSoC drive by users, so we created an intermediary [Contributors Team](https://github.com/orgs/plone/teams/contributors).

Details of this decision are at https://community.plone.org/t/gsoc-students-please-focus-on-your-application-restrictions-on-developer-team-access-on-github/17041

It's still not perfect, because we lack the full GitHub Enterprise plan, but it is the best we can manage for an open source project plan on GitHub.

MrTango reviewed (commented):
looks good in genneral

stevepiercy committed:
Merge branch '6.0' into first-time-contributor-update

stevepiercy committed:
Add section "Don't create and close multiple pull requests for the same issue"

Why is this necessary? See:
- https://github.com/plone/volto/pulls?q=is%3Apr+author%3ADhanushShettyH
- https://github.com/plone/volto/pulls?q=is%3Apr+author%3Asah246ir
- https://github.com/plone/volto/pulls?q=is%3Apr+author%3AJASIM0021
- https://github.com/plone/volto/pulls?q=is%3Apr+author%3Aabhinaisai2002
- https://github.com/plone/volto/pulls?q=is%3Apr+author%3Arabahalishah
- https://github.com/plone/volto/pulls?q=is%3Apr+author%3Amehedikhan72

stevepiercy committed:
Merge remote-tracking branch 'origin/first-time-contributor-update' into first-time-contributor-update

stevepiercy committed:
Merge branch '6.0' into first-time-contributor-update

# Conflicts:
#	docs/contributing/first-time.md

ichim-david commented:
@stevepiercy I still don't see any mention of how to run linters and tests which most of the times they need to run in order to fix their commits.

stevepiercy commented:
> @stevepiercy I still don't see any mention of how to run linters and tests which most of the times they need to run in order to fix their commits.

Because each project has its own specific `make` commands, they should be documented in each project. For example, in Volot, see https://github.com/plone/volto/pull/5377/files.

But that's just documentation. I agree that the content experts in each project need to document how to contribute, as if they are first-timers. This is especially needed in Volto.

stevepiercy committed:
Merge branch '6.0' into first-time-contributor-update

stevepiercy committed:
Merge branch '6.0' into first-time-contributor-update

stevepiercy commented:
This has been marinating long enough for people to provide feedback since the Steering Circle meeting last week. I'll merge as soon as it is green.
